wit/types.wit

```diff
@@ -1,44 +1,77 @@
-        channels: list<channel>,
-        extensions: option<list<tuple<string, string>>>
+    /// Configuration includes a required list of topics the guest is subscribing to
+    record config {
+        topics: list<topic>,
     }
   
-    /// Format specification for messages 
-    ///  - more info: https://github.com/clemensv/spec/blob/registry-extensions/registry/spec.md#message-formats
-    ///  - message metadata can further decorate w/ things like format version, and so on.
-    enum format-spec {
-        cloudevents,
-        http,
-        amqp,
-        mqtt,
-        kafka,
-        raw
-    }
-  
-    /// A message with a binary payload, a format specification, and decorative metadata.
-    record message {
-        data: list<u8>,
-        format: format-spec,
-        metadata: option<list<tuple<string, string>>>
+    /// A message with a binary payload and additional information
+    resource message {
+        constructor(topic: topic, data: list<u8>, content-type: option<string>, metadata: option<list<tuple<string, string>>>);
+        /// The topic/subject/channel this message was received or should be sent on
+        topic: func() -> topic;
+        /// Set the topic/subject/channel this message should be sent on
+        set-topic: func(topic: topic);
+        /// An optional content-type describing the format of the data in the message. This is 
+        /// sometimes described as the "format" type
+        content-type: func() -> option<string>;
+        /// Set the content-type describing the format of the data in the message. This is
+        /// sometimes described as the "format" type
+        set-content-type: func(content-type: string);
+        /// An opaque blob of data
+        data: func() -> list<u8>;
+        /// Set the opaque blob of data for this message, discarding the old value
+        set-data: func(data: list<u8>);
+        /// Optional metadata (also called headers or attributes in some systems) attached to the
+        /// message
+        metadata: func() -> option<list<tuple<string, string>>>;
+        /// Add a new key-value pair to the metadata, overwriting any existing value for the same key
+        add-metadata: func(key: string, value: string);
+
+        /// Completes/acks the message
+        /// 
+        /// A message can exist under several statuses:
+        /// (1) available: the message is ready to be read,
+        /// (2) acquired: the message has been sent to a consumer (but still exists in the queue),
+        /// (3) accepted (result of complete): the message has been received and ACK-ed by a consumer and can be safely removed from the queue,
+        /// (4) rejected (result of abandon): the message has been received and NACK-ed by a consumer, at which point it can be:
+        ///         - deleted,
+        ///         - sent to a dead-letter queue, or
+        ///         - kept in the queue for further processing.
+        complete: func() -> result<_, error>;
```

Comment:
Given that `handler` now takes a single message, could we remove the `complete`/`abandon` methods from the `message` resource and instead use the return value of `handler` to indicate completion/abandonment?  This has a few advantages:
* it avoids the cross-product of spec-questions of what happens if the message was { completed, abandoned, neither } x and the result of `handler` was { `ok`, `error`, trapped }, all of which otherwise need precise answers
* it avoids questions of how `complete` and `abandon` interact with `request-reply`
* it's more like normal success/failure control flow

Reply:
I'm glad you brought this up. I actually started doing it this way and then was unsure if it was a good idea. Now that it also came from someone else I feel more confident 😅